### ISSUE TITLE
Extend fused attention to Xe-HPG (Arc A-series / Flex Series): SDPA is ~4x off the same device's GEMM rate

### ISSUE THREAD

**Issue by AIVirtuoso** [intel/torch-xpu-ops#4886]
First, thanks for the SDPA work that landed in PyTorch 2.7 and for #2200. The oneDNN attention path is a real improvement over what the XPU backend had before, and having Arc listed as first-class supported hardware here has made this card genuinely usable for local diffusion work.

I'd like to make a case that extending the fused/flash attention path to **Xe-HPG** deserves a slot on the roadmap, with measurements rather than just a request.

On an Arc A770, attention is the only operation not keeping up with the silicon:

| operation (bf16) | achieved |
|---|---|
| `F.linear` `[8112,2048] @ [2048,8192]` | **91.9 TFLOPS** |
| `F.linear` `[8112,2048] @ [2048,2048]` | **98.8 TFLOPS** |
| `scaled_dot_product_attention` (2x16x4056x128) | **21.7 TFLOPS** |

Same device, same dtype, same run. The GEMM figures show the hardware is healthy and the oneDNN matmul path is well tuned; attention sits at roughly a quarter of that.

A second data point suggests the current path may not be fused in the usual sense: a hand-written attention that **materialises the entire SxS score matrix** reaches 16.1 TFLOPS, only 1.35x behind the built-in path. A fused implementation would normally beat a materialised one by considerably more.

Backend availability on `2.14.0.dev20260808+xpu`:

```
  default (whatever torch picks)        12.44 ms    21.7 TFLOPS
  MATH                                  58.75 ms     4.6 TFLOPS
  FLASH_ATTENTION                    unavailable (RuntimeError)
  EFFICIENT_ATTENTION                   58.77 ms     4.6 TFLOPS   ("not supported on XPU, falling back to math")
  OVERRIDEABLE                          12.40 ms    21.7 TFLOPS
  F.linear reference                     2.96 ms    91.9 TFLOPS
```

**Why this may be worth more than one user's convenience.** The Data Center GPU **Flex Series is the same Xe-HPG silicon**, not merely a related architecture:

```
56a0  DG2 [Arc A770]                        <- this report
56c0  ATS-M [Data Center GPU Flex 170]
56c1  ATS-M [Data Center GPU Flex 140]
56c2  ATS-M [Data Center GPU Flex 170V]
```

(device IDs from [`hwdata`'s `pci.ids`](https://github.com/vcrhonek/hwdata/blob/master/pci.ids), the database `lspci` uses.)

The Flex 170 is ACM-G10 with 32 Xe cores and 16 GB, the same die and configuration as the A770, and Intel documents the Flex Series as ["leveraging Xe-HPG microarchitecture"](https://www.intel.com/content/www/us/en/products/details/discrete-gpus/data-center-gpu/flex-series/products.html) ([Flex 170 = ACM-G10, Flex 140 = dual ACM-G11](https://www.tomshardware.com/news/intel-launches-data-center-gpu-flex-series)). Both Arc A770 and the Flex Series are listed as supported hardware here. So a fused attention kernel targeting Xe-HPG would land on a product line sold into inference deployments today, not only on 2022 consumer parts; the work looks shared rather than consumer-only.

**Attention is where the remaining headroom is.** Profiling a 2B-parameter diffusion transformer (28 blocks, hidden 2048, MLP 8192) at 1 MP:

```
op (per block)                                 ms    TFLOPS
self_attn qkv      [M,2048]@[2048,6144]     2.307      88.5
self_attn out      [M,2048]@[2048,2048]     0.689      98.8
mlp up             [M,2048]@[2048,8192]     2.958      92.0
mlp down           [M,8192]@[8192,2048]     5.184      52.5
self_attn SDPA (4056 tokens)               12.371      21.8   <-- 46% of the block
cross_attn SDPA (ctx 256)                   0.887      19.2
```

Attention is **46% of total step time at ~22% of the device's demonstrated rate**. Reaching even 60 TFLOPS would cut end-to-end image generation by roughly 25% on this hardware.

Related issues: #2200 (support flash attention op on XPU device, closed at milestone PT2.13, filed against a B60; the backend is still unavailable on A-series), #2390 (SDPA uses a different backend than IPEX), #2232 (SDPA backward kernel).

**Questions**

1. Is the #2200 flash-attention work architecture-gated, and is Xe-HPG in or out of scope? Given the Flex Series shares the architecture, including it may be cheaper than it first appears.
2. Is `EFFICIENT_ATTENTION` on XPU planned, or is the oneDNN path intended to be the only implementation?
3. If Xe-HPG is out of scope, an explicit statement would genuinely help; a clear answer is more useful than optimistic guessing, and I'd rather plan around it than keep testing.

Happy to test pre-release builds, run this benchmark across other shapes or dtypes, or provide profiles from real diffusion workloads if useful.

### Alternatives

I worked through the framework-level options before filing, and none of them address this:

| approach | result on A770 |
|---|---|
| `torch.compile` (Inductor) | **1.01x**: GEMMs still dispatch to oneDNN, and there is little elementwise work to fuse |
| OpenVINO as a `torch.compile` backend | **1.03x** on a fully captured block (0 graph breaks). Expected, since `libopenvino_intel_gpu_plugin.so` links oneDNN, so it calls the same kernels |
| INT8 via `torch._int_mm` | The GEMM itself is **2x faster than bf16** (126.9 TOPS), but dynamic quantisation's dequant epilogue costs more than the GEMM saves; net slower end to end |
| `SDPBackend.MATH` | 4.6 TFLOPS, ~4.7x slower than the default path |
| `SDPBackend.EFFICIENT_ATTENTION` | Falls back to math with *"not supported on XPU"* |
| `--force-channels-last`, non-blocking transfers, memory-format changes | Sub-1% each |

The GEMM path is in good shape; a faster attention kernel appears to be the only remaining lever on this architecture.

### Additional context

**Reproducer** (self-contained, no models required):

```python
import time
import torch
import torch.nn.functional as F
from torch.nn.attention import SDPBackend, sdpa_kernel

DEV, DT = "xpu:0", torch.bfloat16
B, H, S, D = 2, 16, 4056, 128
FLOPS = 4 * B * H * S * S * D

def bench(fn, n=10):
    for _ in range(3): fn()
    torch.xpu.synchronize()
    t0 = time.perf_counter()
    for _ in range(n): fn()
    torch.xpu.synchronize()
    return (time.perf_counter() - t0) / n

q, k, v = (torch.randn(B, H, S, D, device=DEV, dtype=DT) for _ in range(3))
print(torch.__version__, torch.xpu.get_device_name(0))

t = bench(lambda: F.scaled_dot_product_attention(q, k, v))
print(f"default             {t*1e3:7.2f} ms  {FLOPS/t/1e12:6.1f} TFLOPS")

for name in ("MATH", "FLASH_ATTENTION", "EFFICIENT_ATTENTION", "OVERRIDEABLE"):
    b = getattr(SDPBackend, name, None)
    if b is None:
        print(f"{name:<20} not in this build"); continue
    try:
        with sdpa_kernel(b):
            t = bench(lambda: F.scaled_dot_product_attention(q, k, v))
        print(f"{name:<20}{t*1e3:7.2f} ms  {FLOPS/t/1e12:6.1f} TFLOPS")
    except Exception as e:
        print(f"{name:<20} unavailable ({type(e).__name__})")

a = torch.randn(B*S, 2048, device=DEV, dtype=DT)
w = torch.randn(8192, 2048, device=DEV, dtype=DT)
t = bench(lambda: F.linear(a, w))
print(f"F.linear reference  {t*1e3:7.2f} ms  "
      f"{2*B*S*2048*8192/t/1e12:6.1f} TFLOPS")
```

**Environment**

| | |
|---|---|
| GPU | Intel Arc A770 16 GB (`8086:56a0`, DG2-G10 C0, Xe-HPG) |
| Kernel driver | `xe` (Linux 7.1.7), GuC 70.53.0, DMC 2.8 |
| Compute runtime | intel-compute-runtime 26.18.38308.1 |
| Level Zero | loader 1.28.5 |
| torch | `2.14.0.dev20260808+xpu`; also reproduced on `2.13.0+xpu` |

**Comment by AIVirtuoso** [comment]
Closing this myself — I found the answer, and two of the claims above were wrong. Writing it up rather than deleting, in case anyone else lands here.

**Q1, answered from this repo's source: Xe-HPG is out of scope, and cannot reasonably be brought in.** The SYCL-TLA flash implementation is closed to it three times over:

- the runtime gate lists exactly `intel_gpu_pvc`, `intel_gpu_pvc_vg`, `intel_gpu_bmg_g21`, `intel_gpu_bmg_g31` (`mha_fwd.cpp:104-116`) — this is the literal source of the `RuntimeError` from `SDPBackend.FLASH_ATTENTION` I reported;
- the mainloop is built from `make_block_2d_copy_A/B` (`collective/xe_fmha_fwd_mainloop.h:106-117`);
- it compiles with `-spirv-ext=+SPV_INTEL_2d_block_io` and AOT target `-device pvc,bmg`, where the rest of the repo targets `pvc,bmg,dg2,arl-h,mtl-h,lnl-m,ptl-h,ptl-u` (`cmake/BuildFlags.cmake:202-216`).

The A770 reports `has_subgroup_2d_block_io: False`, so this isn't a scoping decision that could be revisited — the implementation is built on a hardware capability the part doesn't have. My Flex Series argument doesn't rescue it either: ATS-M is the same silicon and therefore has the same gap. An explicit "out of scope" is what I asked for in Q3, and the source gives it more clearly than a maintainer reply could have.

**Correction 1: the existing path is already fused.** I inferred it might not be, from a materialised implementation coming within 1.35x. That inference was wrong — `ONEDNN_VERBOSE=1` shows `ocl:micro:reusable` dispatched at every head_dim, and it is a proper flash-style kernel with online softmax that never materialises the S×S score matrix. The materialised baseline is dominated by the D-independent score matrix, so its cost happens to land near the fused kernel's at D=128. Coincidence, not evidence.

**Correction 2: the 21.7 TFLOPS was a config-table entry, not a missing kernel.** oneDNN picks a tiling from a static table keyed on `[arch, head_size, seq_len, properties]`. The Xe-HPG `head_size=128` entry uses a 32-subgroup workgroup at 34.2 KiB SLM; a 16-subgroup one at 25.2 KiB is **1.6-1.8x faster** on this card, because two of the smaller workgroups fit per Xe core and can overlap each other's barriers. Measured at `1x16x4056x128`: 6.65 ms to 3.74 ms, correctness-checked with `benchdnn --mode=C`. I swept all 74 distinct valid tilings and that one is optimal. Filing it against oneDNN, since that is where the table lives.

**Correction 3: my "reaching even 60 TFLOPS" premise was unachievable.** After retuning, attention is at the ceiling its *shapes* allow. Running the same two matmuls as a plain `bmm` — no softmax, no fusion — takes exactly as long:

| | time | TFLOPS |
|---|---|---|
| QK^T as `bmm` (K=128) | 1.89 ms | 35.6 |
| PV as `bmm` (N=128) | 1.86 ms | 36.3 |
| both matmuls alone | 3.75 ms | 35.9 |
| fused SDPA, softmax included | 3.75 ms | 35.6 |

`head_dim` is the thin dimension in both matmuls, and this part needs K or N >= 512 to approach peak (measured: K=128 gives 27.9 TFLOPS, K=1024 gives 64.6, K=4096 gives 81.2). So no attention implementation, flash or otherwise, gets past ~36 TFLOPS at head_dim 128 here. The gap I reported was real; my explanation of it was not.

The GEMM numbers in the original report stand, and so does the practical point that attention dominates diffusion step time on this hardware — it just turned out to be a tuning issue with a much smaller ceiling than I assumed, and the fix belongs in **uxlfoundation/oneDNN** (this repo vendors no oneDNN, per `README.md:21` — the operators live in pytorch core under `aten/src/ATen/native/mkldnn/xpu`).

Thanks for the SDPA work regardless; the oneDNN path is what made this card usable, and the remaining 1.6x was sitting in one table row.

Written with the help of an LLM.
